pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,341 @@
+	cache kubecontainer.Cache
+	// For testability.
+	clock clock.Clock
+
+	// GenericPLEG is used to force relist when required
+	genericPleg PodLifecycleEventGenerator
+
+	// The maximum number of retries when getting container events from the runtime.
+	eventedPlegMaxStreamRetries int
+
+	// The period for relisting all containers using generic pleg as a fallback
+	genericPlegRelistPeriod time.Duration
+}
+
+// NewEventedPLEG instantiates a new EventedPLEG object and return it.
+func NewEventedPLEG(runtime kubecontainer.Runtime, runtimeService internalapi.RuntimeService, eventChannel chan *PodLifecycleEvent,
+	cache kubecontainer.Cache, genericPleg PodLifecycleEventGenerator, eventedPlegMaxStreamRetries int,
+	genericPlegRelistPeriod time.Duration, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:                     runtime,
+		runtimeService:              runtimeService,
+		eventChannel:                eventChannel,
+		cache:                       cache,
+		genericPleg:                 genericPleg,
+		eventedPlegMaxStreamRetries: eventedPlegMaxStreamRetries,
+		genericPlegRelistPeriod:     genericPlegRelistPeriod,
+		clock:                       clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent events.
+// TODO: support multiple subscribers.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Relist relists all containers using GenericPLEG
+func (e *EventedPLEG) Relist() {
+	e.genericPleg.Relist()
+}
+
+// Start spawns a goroutine to relist periodically.
+func (e *EventedPLEG) Start() {
+	IsEventedPLEGInUse = true
+	go wait.Forever(e.watchEventsChannel, 0)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
+		return false, fmt.Errorf("EventedPLEG: pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+
+	timestamp := e.clock.Now()
+	metrics.PLEGLastSeen.Set(float64(timestamp.Unix()))
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go func() {
+		numAttempts := 0
+		for {
+			err := e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+			if err == io.EOF {
+				// The runtime service has been closed, exit the goroutine.
```

Comment:
This is where we could potentially signal Not Healthy.

Reply:
As of now, we don't use the health information of evented PLEG anywhere [1] (unlike generic PLEG: https://github.com/kubernetes/kubernetes/blob/d6a3a68afce1525dc31a2a8e73a2bc5903120eb9/pkg/kubelet/kubelet.go#L688 which uses health info to mark node state). Essentially, it's no-op rn to just help us implement the `PodLifecycleEventGenerator` interface; but if absolutely necessary we could add another variable tracking the state of this condition and checking that inside `Healthy()`.

Also, we wouldn't want to add the healthy call of evented PLEG (like above) as evented PLEG falls back to generic PLEG-like behaviour when number of connection attempts to evented CRI is exhausted.

[1] [111384 comment](https://github.com/kubernetes/kubernetes/pull/111384/#discussion_r935547740)

Reply:
We should be reporting healthy on evented pleg as soon as possible like generic pleg - since it's more likely we'll find conditions that are problematic in evented pleg during the soak period.